src/optimizer/expr_simplification.ts

```diff
@@ -0,0 +1,713 @@
+                        }
+                    }
+                }
+
+                newAst = cloneAstNode({
+                    ...t.ast,
+                    declarations: newDeclarations,
+                });
+
+                break;
+            }
+            case "contract": {
+                const newDeclarations: AstContractDeclaration[] = [];
+
+                for (const decl of t.ast.declarations) {
+                    switch (decl.kind) {
+                        case "asm_function_def":
+                        case "function_def": {
+                            const newCode = newMethods.get(idText(decl.name))!
+                                .ast as AstContractDeclaration;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "constant_def": {
+                            const newCode = newConstantsMap.get(
+                                idText(decl.name),
+                            )!.ast as AstContractDeclaration;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "field_decl": {
+                            const newCode = newFieldsMap.get(
+                                idText(decl.name),
+                            )!.ast;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "receiver": {
+                            const newCode = newReceiversMap.get(decl.id)!.ast;
+                            newDeclarations.push(newCode);
+                            break;
+                        }
+                        case "contract_init":
+                            newDeclarations.push(newInitializer!.ast);
+                            break;
+                    }
+                }
+
+                newAst = cloneAstNode({
+                    ...t.ast,
+                    declarations: newDeclarations,
+                });
+
+                break;
+            }
+        }
+
+        newTypes.set(t.name, {
+            ...t,
+            ast: newAst,
+            init: newInitializer,
+            constants: newConstants,
+            fields: newFields,
+            functions: newMethods,
+            receivers: newReceivers,
+        });
+    }
+    ctx = replaceTypes(ctx, newTypes);
+
+    return ctx;
+}
+
+function process_statements(
+    statements: AstStatement[],
+    ctx: CompilerContext,
+    interpreter: Interpreter,
+): { stmts: AstStatement[]; ctx: CompilerContext } {
+    const newStatements: AstStatement[] = [];
+
+    for (const stmt of statements) {
+        const result = process_statement(stmt, ctx, interpreter);
+        newStatements.push(result.stmt);
+        ctx = result.ctx;
+    }
+
+    return { stmts: newStatements, ctx: ctx };
+}
+
+function process_statement(
```

Comment:
This is a visitor, it's not unique to optimizer, and likely should go into `ast.ts`.